66_HLSLBxDFTests/app_resources/test_components.hlsl

```diff
@@ -0,0 +1,378 @@
+            transmitted = cache.isTransmission();
+        }
+        else
+        {
+            dg1_query_type dq = base_t::bxdf.ndf.template createDG1Query<iso_interaction, iso_cache>(base_t::isointer, isocache);
+            fresnel_type _f = base_t::bxdf_t::__getOrientedFresnel(base_t::bxdf.fresnel, base_t::isointer.getNdotV());
+            quant_query_type qq = bxdf::impl::quant_query_helper<ndf_type, fresnel_type, base_t::bxdf_t::IsBSDF>::template __call<iso_interaction, iso_cache>(base_t::bxdf.ndf, _f, base_t::isointer, isocache);
+            quant_type DG1 = base_t::bxdf.ndf.template DG1<sample_t, iso_interaction>(dq, qq, s, base_t::isointer, isNdfInfinity);
+            dg1 = DG1.microfacetMeasure * hlsl::abs(isocache.getVdotH() / base_t::isointer.getNdotV());
+            reflectance = _f(isocache.getVdotH())[0];
+            NdotH = isocache.getAbsNdotH();
+            transmitted = isocache.isTransmission();
+        }
+
+        if (isNdfInfinity)
+            return BET_INVALID;
+
+        if (transmitted)
+        {
+            float eta = base_t::rc.eta.x;
+            if (base_t::isointer.getNdotV() < 0.f)
+                eta = 1.f / eta;
+
+            reflectance = transmitted ? 1.f - reflectance : reflectance;
+            NBL_IF_CONSTEXPR(aniso)
+            {
+                float denom = cache.getVdotH() + eta * cache.getLdotH();
+                dg1 =  dg1 * hlsl::abs(eta * eta * cache.getLdotH()) / (denom * denom);
+            }
+            else
+            {
+                float denom = isocache.getVdotH() + eta * isocache.getLdotH();
+                dg1 = dg1 * hlsl::abs(eta * eta * isocache.getLdotH()) / (denom * denom);
+            }
+        }
+        else
+        {
+            NBL_IF_CONSTEXPR(aniso)
+                dg1 = 0.25f * dg1 / hlsl::abs(cache.getVdotH());
+            else
+                dg1 = 0.25f * dg1 / hlsl::abs(isocache.getVdotH());
+        }
+
+        return BET_NONE;
+    }
+
+    ErrorType test()
+    {
+        if (traits_t::type == bxdf::BT_BRDF)
+        {    
+            if (base_t::isointer.getNdotV() <= bit_cast<float>(numeric_limits<float>::min))
+                return BET_INVALID;
+        }        
+        else if (traits_t::type == bxdf::BT_BSDF)
+        {
+            if (hlsl::abs(base_t::isointer.getNdotV()) <= bit_cast<float>(numeric_limits<float>::min))
+                return BET_INVALID;
+        }
+
+        ErrorType res = compute();
+        if (res != BET_NONE)
+            return res;
+
+        // get jacobian
+        float32_t2x2 m = float32_t2x2(
+            sx.getTdotL() - s.getTdotL(), sy.getTdotL() - s.getTdotL(),
+            sx.getBdotL() - s.getBdotL(), sy.getBdotL() - s.getBdotL()
+        );
+        float det = nbl::hlsl::determinant<float32_t2x2>(m) / (eps * eps);
+        
+        float jacobi_dg1_ndoth = det * dg1 / hlsl::abs(s.getNdotL());
+        if (!checkZero<float>(jacobi_dg1_ndoth - 1.f, 5e-2))
```

Comment:
if you get 0 assigned to `alpha` how do you protect against false test negatives ?
(0 roughness NDF lobes will produce zero measure jacobians)